src/foam/nanos/auth/services.jrl

```diff
@@ -459,3 +483,73 @@ p({
     }
   """
 })
+
+p({
+  "class": "foam.nanos.boot.NSpec",
+  "name": "userRegistrationDAO",
+  "lazy": false,
+  "serve": true,
+  "authenticate": false,
+  "authNotes": "UserRegistrationDAO disables find, select and remove operations.",
+  "serviceScript":
+  """
+    return new foam.dao.EasyDAO.Builder(x)
+      .setOf(foam.nanos.auth.User.getOwnClassInfo())
+      .setInnerDAO(x.get("localUserDAO"))
+      .setDecorator(
+        new foam.nanos.auth.UserRegistrationSanitationDAO(x,
+        new foam.nanos.auth.UserRegistrationDAO(x,
+        new foam.nanos.auth.UserPasswordHashingDAO(x,
+        new foam.dao.ValidatingDAO(x,
+          new foam.dao.NullDAO(x,foam.nanos.auth.User.getOwnClassInfo()),
+          new foam.nanos.auth.validators.UserGroupAndSpidValidator()
+        )))))
+      .setAuthorize(false)
+      .setRuler(true)
+      .setPm(true)
+      .setValidated(true)
+      .setEnableInterfaceDecorators(false)
+      .build();
+  """,
+  "client": """
+    {
+      "of": "foam.nanos.auth.User",
+      "cache": true,
+      "retryBoxMaxAttempts": 0,
+      "decorators": [
+        { "class": "foam.nanos.fs.FileDAODecorator" },
+        { "class": "foam.nanos.fs.FileArrayDAODecorator"}
+      ];
+    }
+  """
+})
+
+// only available to admin to update passwords.
```

Comment:
The comment is wrong, this is also exposed to the basicUser group.

Overall this service feels unnecessary, can't we use property level permissions to let users change their own password (and let admin change anything?)

Reply:
Changing one's own password uses a DAO stack that has the PasswordHashingDAO. 
When admin, and changing another users password, there does not exist an nspec which to submit to to invoke the hashing of the desired password. 
if this is granted to basicUser then that is a mistake, ther service.daoname permission should not be granted to anyone.  
So open to ideas on how to allow administrators to change another users password.  The current change password views require either email or token verification, or the recent 'insecure' system which only allows change of the subject/context (already logged in users) password.

Reply:
Here you expose service.userUpdatePasswordDAO to "basicUser"

https://github.com/kgrgreer/foam3/pull/3579/files/ce70ed0438e8fb343c4569ab2f67a82aa6de237f#diff-13acd8f684fcd591af01935a858aff229a504b34ee3ca83176f9118144228f46R41

Reply:
this feels flaky though, can't we just decorate UserDAO in all cases to hash the password if the password field changes?  Weird that we have to expose the userDAO as these other special services to hash the passwords.

Reply:
Services and permissions fixed. 
We have pre-existing password protection services with email and token verification, which rely on verifying secondary information before allowing a password update.  These are onerous for a simple foam application, but also preclude placing the UserPasswordHashingDAO on userDAO.  The resetPasswordInsecureService was added for the simple app, with the only protection being that one can only update the logged in user password (your own).   Also, hitherto, it was not possible for admin to update another users password outside of using a script.  So the non-permissioned userUpdatePasswordDAO was added which allows both the user id and password to be provided. 
I agree that this seems a little much for password management, and I was surprised by lack of support the admin and the simple case.  I'll put more thought into a better system, and open to your ideas.  But for the moment this gets us status-quo.  The user only interacts via the Change Password menu, so we have lots of leeway to change the backend.

Reply:
Sounds good, thanks for responding to my comments, I'm also still getting up to speed on how auth has evolved in nanos.

Reply:
Ok, previous to the clientLoginService PR, things worked more like you expected. 
I've reverted to that setup, allowing removal of nspec additions for userUpdatePassword for admin, and the PasswordResetInsecureService for users. 
An application has full control to change the userDAO stack, which nanopay does (for example). 
And if an application allows access to the User model by a User, then a user could update password directly from the desiredPassword field.